llvm/tools/repo-fragments/RepoFragments.cpp

```diff
@@ -141,8 +141,9 @@ int main(int argc, char *argv[]) {
   };
 
   const auto *Sep = "";
-  for (auto const &CM :
-       *pstore::repo::compilation::load(Db, CompilationPos->second)) {
+  auto const Compilation =
```

Comment:
`Compilation` variable is only used one time, why need  it?

Reply:
We crash if it’s not there.

Most of the time, the `shared_ptr<>` returned from loading data from the store will point into memory-mapped storage with the lifetime of the underlying `pstore::database` instance. Sometimes, however, you’ll encounter a _spanning_ load which crosses the boundary between two memory-mapped regions. In this case, the lifetime of the memory is that of the `shared_ptr<>` itself.

In the original version, the shared_ptr is discarded after it is dereferenced to yield the for-loop range. As soon as we try to access its contents inside the loop, the memory has gone and we crash. The solution is to ensure that we maintain a reference to the shared_ptr for the duration of the loop.

Uncovering these bugs is the job of the PSTORE_ALWAYS_SPANNING configuration variable.